# Exploratory Data Analysis Report

# Data Loading - 2023 US Civil Flights

In [5]:
import os
import pandas as pd
import kagglehub

# Download dataset via kagglehub
dataset_path = kagglehub.dataset_download(
    "bordanova/2023-us-civil-flights-delay-meteo-and-aircraft"
)

print("Dataset downloaded to:")
print(dataset_path)

# List CSV files
csv_files = [f for f in os.listdir(dataset_path) if f.endswith(".csv")]

print("\nCSV files found:")
for f in csv_files:
    print(f"- {f}")

# Quick overview (same EDA logic as your original code)
for file in csv_files:
    try:
        df = pd.read_csv(os.path.join(dataset_path, file))

        print(f"\n===== QUICK OVERVIEW FOR {file} =====\n")
        print(f"Shape: {df.shape[0]} rows x {df.shape[1]} columns\n")

        print("Columns:")
        for col in df.columns:
            print(f"- {col}")

        print("\nInfo:\n")
        print(df.info(), "\n")

        print("First 5 rows:\n")
        print(df.head(), "\n")

        print("===== END OF QUICK OVERVIEW =====\n")

    except Exception as e:
        print(f"Error while processing {file}: {e}")


100%|██████████| 127M/127M [00:06<00:00, 21.0MB/s] 

Extracting files...


Dataset downloaded to:
C:\Users\flori\.cache\kagglehub\datasets\bordanova\2023-us-civil-flights-delay-meteo-and-aircraft\versions\9

CSV files found:
- airports_geolocation.csv
- Cancelled_Diverted_2023.csv
- maj us flight - january 2024.csv
- US_flights_2023.csv
- weather_meteo_by_airport.csv

===== QUICK OVERVIEW FOR airports_geolocation.csv =====

Shape: 364 rows x 7 columns

Columns:
- IATA_CODE
- AIRPORT
- CITY
- STATE
- COUNTRY
- LATITUDE
- LONGITUDE

Info:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 364 entries, 0 to 363
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   IATA_CODE  364 non-null    object 
 1   AIRPORT    364 non-null    object 
 2   CITY       364 non-null    object 
 3   STATE      364 non-null    object 
 4   COUNTRY    364 non-null    object 
 5   LATITUDE   364 non-null    float64
 6   LONGITUDE  364 non-null    float64
dtypes: float64(2), object(5)
memory usage: 20.0+ KB
None 

First 5

# Data Description 

The dataset contains multiple CSV files that provide complementary information about US civil flights. The main file, US_flights_2023.csv, records over 6.7 million flights with details on departure and arrival times, delays, aircraft information, and distances. 

Cancelled_Diverted_2023.csv focuses on flight cancellations and diversions, with just over 100,000 entries. 

maj us flight - january 2024.csv contains approximately half a million flights from January 2024, including aircraft manufacturer and age. 

airports_geolocation.csv provides latitude and longitude for 364 airports, along with city and state information. 

Finally, weather_meteo_by_airport.csv contains daily weather measurements for airports, including temperature, precipitation, wind, and pressure. 

In [7]:
import os
import pandas as pd
import numpy as np
import kagglehub

# List CSV files
csv_files = [f for f in os.listdir(dataset_path) if f.endswith(".csv")]

print("\nCSV files found:")
for f in csv_files:
    print(f"- {f}")

def dataset_summary(df):
    print("===== DATASET SUMMARY =====\n")
    print(f"Shape: {df.shape[0]} rows x {df.shape[1]} columns\n")
    
    # Missing values
    missing_count = df.isna().sum()
    missing_percent = (missing_count / len(df) * 100).round(2)
    missing_report = pd.DataFrame({
        'Missing Count': missing_count,
        'Missing %': missing_percent
    }).sort_values('Missing Count', ascending=False)
    print(">>> Missing Values:\n", missing_report, "\n")
    
    # Duplicate rows
    dup_count = df.duplicated().sum()
    print(f">>> Duplicate Rows: {dup_count}\n")
    
    # Data types
    print(">>> Data Types:\n", df.dtypes, "\n")
    
    # Basic statistics (numerical columns)
    if df.select_dtypes(include=np.number).shape[1] > 0:
        print(">>> Basic Statistics (numerical columns):\n", df.describe().T, "\n")
    
    # Basic statistics (categorical columns)
    if df.select_dtypes(include='object').shape[1] > 0:
        print(">>> Basic Statistics (categorical columns):\n", df.describe(include='object').T, "\n")
    
    print("===== END OF REPORT =====\n\n")

# Iterate over each CSV file to generate summary
for file in csv_files:
    try:
        df = pd.read_csv(os.path.join(dataset_path, file))
        print(f"\n===== SUMMARY FOR {file} =====\n")
        dataset_summary(df)
    except Exception as e:
        print(f"Error while processing {file}: {e}\n")



CSV files found:
- airports_geolocation.csv
- Cancelled_Diverted_2023.csv
- maj us flight - january 2024.csv
- US_flights_2023.csv
- weather_meteo_by_airport.csv

===== SUMMARY FOR airports_geolocation.csv =====

===== DATASET SUMMARY =====

Shape: 364 rows x 7 columns

>>> Missing Values:
            Missing Count  Missing %
IATA_CODE              0        0.0
AIRPORT                0        0.0
CITY                   0        0.0
STATE                  0        0.0
COUNTRY                0        0.0
LATITUDE               0        0.0
LONGITUDE              0        0.0 

>>> Duplicate Rows: 0

>>> Data Types:
 IATA_CODE     object
AIRPORT       object
CITY          object
STATE         object
COUNTRY       object
LATITUDE     float64
LONGITUDE    float64
dtype: object 

>>> Basic Statistics (numerical columns):
            count       mean        std        min         25%       50%  \
LATITUDE   364.0  38.810970   8.757505  -14.27810   33.672655  39.25193   
LONGITUDE  364.0 -96.

# Key Metrics from each File

airports_geolocation.csv has 364 complete rows with location data and no duplicates. 

Cancelled_Diverted_2023.csv has 104,488 rows, no missing values, 945 duplicates, mostly cancelled flights, skewed delay distributions, 15 airlines, and ~345 airports. 

maj us flight - january 2024.csv contains 527,197 rows, no missing values, three duplicates, skewed delay metrics, and aircraft details (manufacturer, model, age ~14 years). 

US_flights_2023.csv has 6,743,404 rows, no missing values, 31 duplicates, similar delay distributions, and consistent categorical data. 

weather_meteo_by_airport.csv has 132,860 rows, no missing values, and complete weather metrics per airport

In the flight datasets (Cancelled_Diverted_2023.csv, maj us flight - january 2024.csv, US_flights_2023.csv), median departure and arrival delays are close to zero (or negative in January 2024), while maximum delays reach thousands of minutes (e.g., 3,024–4,413 minutes). This suggests that most flights have minimal delay, but a small number of flights experience extreme delays.

Delay components broken down by cause—
carrier, 
weather, 
NAS, 
security, 
last aircraft
—show similar patterns: mostly zero or very low values, with occasional extreme spikes. 

In [9]:
import os
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

flights_2023 = pd.read_csv(os.path.join(dataset_path, "US_flights_2023.csv"))
flights_jan2024 = pd.read_csv(os.path.join(dataset_path, "maj us flight - january 2024.csv"))
cancelled_diverted = pd.read_csv(os.path.join(dataset_path, "Cancelled_Diverted_2023.csv"))
airports = pd.read_csv(os.path.join(dataset_path, "airports_geolocation.csv"))
weather = pd.read_csv(os.path.join(dataset_path, "weather_meteo_by_airport.csv"))

for df in [flights_2023, flights_jan2024, cancelled_diverted]:
    df['FlightDate'] = pd.to_datetime(df['FlightDate'])
weather['time'] = pd.to_datetime(weather['time'])

flights_2023.rename(columns={'Aicraft_age':'Aircraft_age'}, inplace=True)
flights_jan2024.rename(columns={'Aicraft_age':'Aircraft_age'}, inplace=True)

flights = pd.concat([flights_2023, flights_jan2024], ignore_index=True)

airports_subset = airports[['IATA_CODE','LATITUDE','LONGITUDE']]
flights = flights.merge(airports_subset, left_on='Dep_Airport', right_on='IATA_CODE', how='left')
flights.rename(columns={'LATITUDE':'Dep_LAT','LONGITUDE':'Dep_LON'}, inplace=True)
flights.drop(columns=['IATA_CODE'], inplace=True)

flights = flights.merge(airports_subset, left_on='Arr_Airport', right_on='IATA_CODE', how='left')
flights.rename(columns={'LATITUDE':'Arr_LAT','LONGITUDE':'Arr_LON'}, inplace=True)
flights.drop(columns=['IATA_CODE'], inplace=True)

weather_agg = weather.groupby('airport_id').agg(
    tavg_dep=('tavg','mean'),
    tmin_dep=('tmin','mean'),
    tmax_dep=('tmax','mean'),
    prcp_dep=('prcp','sum'),
    snow_dep=('snow','sum'),
    wspd_dep=('wspd','mean')
).reset_index()

flights = flights.merge(weather_agg, left_on='Dep_Airport', right_on='airport_id', how='left')
flights.drop(columns=['airport_id'], inplace=True)

delay_cols = ['Dep_Delay','Arr_Delay','Delay_Carrier','Delay_Weather',
              'Delay_NAS','Delay_Security','Delay_LastAircraft']

print("\n=== Delay Summary ===")
print(flights[delay_cols].describe(percentiles=[0.25,0.5,0.75,0.9,0.99]))

for col in delay_cols:
    print(f"{col} skew: {flights[col].skew():.2f}")

median_by_airline = flights.groupby('Airline')['Dep_Delay'].median().sort_values(ascending=False)
print("\nMedian Dep_Delay by Airline:\n", median_by_airline)

median_by_dep_airport = flights.groupby('Dep_Airport')['Dep_Delay'].median().sort_values(ascending=False)
print("\nMedian Dep_Delay by Departure Airport (top 10):\n", median_by_dep_airport.head(10))

weather_cols = ['tavg_dep','tmin_dep','tmax_dep','prcp_dep','snow_dep','wspd_dep']
print("\nCorrelation of weather with Dep_Delay:\n", flights[weather_cols + ['Dep_Delay']].corr()['Dep_Delay'].sort_values(ascending=False))

sns.set(style="whitegrid")

plt.figure(figsize=(10,6))
sns.histplot(flights['Dep_Delay'], bins=200, kde=False)
plt.xlim(-60,300)
plt.xlabel("Departure Delay (min)")
plt.title("Distribution of Departure Delays")
plt.show()

plt.figure(figsize=(12,6))
sns.boxplot(x='Airline', y='Dep_Delay', data=flights)
plt.ylim(-30,300)
plt.xticks(rotation=45)
plt.title("Departure Delay by Airline")
plt.show()

plt.figure(figsize=(10,6))
sns.scatterplot(x='tavg_dep', y='Dep_Delay', alpha=0.1, data=flights)
plt.ylim(-30,300)
plt.xlabel("Average Temperature (°C)")
plt.ylabel("Departure Delay (min)")
plt.title("Temperature vs Departure Delay")
plt.show()

airport_delay_map = flights.groupby(['Dep_Airport','Dep_LAT','Dep_LON'])['Dep_Delay'].median().reset_index()
plt.figure(figsize=(12,6))
sns.scatterplot(x='Dep_LON', y='Dep_LAT', size='Dep_Delay', hue='Dep_Delay',
                data=airport_delay_map, legend=False, sizes=(20,200))
plt.title("Median Departure Delay by Airport Location")
plt.xlabel("Longitude")
plt.ylabel("Latitude")
plt.show()


MemoryError: Unable to allocate 256. KiB for an array with shape (32768,) and data type int64